include/oneapi/dpl/internal/dynamic_selection_traits.h

```diff
@@ -0,0 +1,220 @@
+
+  template<typename T>
+  struct has_get_policy : decltype(has_get_policy_impl<T>(0)) {};
+
+  template<typename DSPolicy, typename Function, typename... Args>
+  auto has_submit_impl(...) -> std::false_type;
+
+  template<typename DSPolicy, typename Function, typename... Args>
+  auto has_submit_impl(int) -> decltype(std::declval<DSPolicy>().submit(std::declval<Function>(), std::declval<Args>()...), std::true_type{});
+
+  template<typename DSPolicy, typename Function, typename... Args>
+  struct has_submit : decltype(has_submit_impl<DSPolicy, Function, Args...>(0)) {};
+
+  template<typename DSPolicy, typename Function, typename... Args>
+  auto has_submit_and_wait_impl(...) -> std::false_type;
+
+  template<typename DSPolicy, typename Function, typename... Args>
+  auto has_submit_and_wait_impl(int) -> decltype(std::declval<DSPolicy>().submit_and_wait(std::declval<Function>(), std::declval<Args>()...), std::true_type{});
+
+  template<typename DSPolicy, typename Function, typename... Args>
+  struct has_submit_and_wait : decltype(has_submit_and_wait_impl<DSPolicy, Function, Args...>(0)) {};
+
+  template<typename DSPolicy, typename SelectionHandle,  typename Function, typename... Args>
+  auto has_submit_and_wait_handle_impl(...) -> std::false_type;
+
+  template<typename DSPolicy, typename SelectionHandle,  typename Function, typename... Args>
+  auto has_submit_and_wait_handle_impl(int) -> decltype(std::declval<DSPolicy>().submit_and_wait(std::declval<SelectionHandle>(),
+                                                                                                 std::declval<Function>(),
+                                                                                                 std::declval<Args>()...), std::true_type{});
+
+  template<typename DSPolicy, typename SelectionHandle, typename Function, typename... Args>
+  struct has_submit_and_wait_handle : decltype(has_submit_and_wait_handle_impl<DSPolicy, SelectionHandle , Function, Args...>(0)) {};
+
+  template<typename S, typename Info>
+  auto has_report_impl(...) -> std::false_type;
+
+  template<typename S, typename Info>
+  auto has_report_impl(int) -> decltype(std::declval<S>().report(std::declval<Info>()), std::true_type{});
+
+  template<typename S, typename Info>
+  struct has_report : decltype(has_report_impl<S,Info>(0)) {};
+
+  template<typename S, typename Info>
+  auto has_report_value_impl(...) -> std::false_type;
+
+  template<typename S, typename Info>
+  auto has_report_value_impl(int) -> decltype(std::declval<S>().report(std::declval<Info>(), 0), std::true_type{});
+
+  template<typename S, typename Info>
+  struct has_report_value : decltype(has_report_value_impl<S,Info>(0)) {};
+
+ } //namespace internal
+
+  constexpr int deferred_initialization = -1;
+
+  // required interfaces
+  template<typename DSPolicy>
+  auto get_resources(DSPolicy&& dp) {
+    return std::forward<DSPolicy>(dp).get_resources();
```

Comment:
As for me it's look like return ```std::forward<...>(...)``` is not required here.
We may simple rewrite this as ```return dp.get_resources();```
If I am correct, please check all another places which looks like this.